app/controllers/application_controller.rb

```diff
@@ -213,7 +211,12 @@ def invalid_auth_token(_exception)
 
   def user_fully_authenticated?
     !reauthn? && user_signed_in? &&
-      two_factor_enabled? && is_fully_authenticated?
+      two_factor_enabled? &&
+      session['warden.user.user.session'] &&
+      !session['warden.user.user.session'].try(
+        :[],
+        TwoFactorAuthenticatable::NEED_AUTHENTICATION,
+      )
```

Comment:
1. I think that `.dig` might be clearer here than `.try(:[], ...`
2. negating the result of a `try` is always dicey.... I don't have super full context on this but I think we might want to split it into two parts, one to check presence and then one to actually negate

with dig:

```suggestion
      !session.dig('warden.user.user.session', TwoFactorAuthenticatable::NEED_AUTHENTICATION)
```

Reply:
`undefined method `dig' for #<ActionDispatch::Request::Session:0x000055af93a843f8>` 🙃

Reply:
bummer!